k8s: Disable k8s event handover to kvstore by default

tgraf opened the pull request:
This is a new feature and primiarly targeted at larger scale. It requires the
operator to be functional. This adds complexity which is only required once k8s
apiserver event handling at scale becomes a concern. Require users to opt-in.

Fixes: #7722

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7755)
<!-- Reviewable:end -->

tgraf committed:
k8s: Disable k8s event handover to kvstore by default

This is a new feature and primiarly targeted at larger scale. It requires the
operator to be functional. This adds complexity which is only required once k8s
apiserver event handling at scale becomes a concern. Require users to opt-in.

Fixes: #7722

Signed-off-by: Thomas Graf <thomas@cilium.io>

ianvernon commented on daemon/k8s_watcher.go:
```diff
@@ -913,22 +920,33 @@ func (d *Daemon) EnableK8sWatcher(queueSize uint) error {
+			// Wait for kvstore to be connected and consider
+			// switching to kvstore based event handling if the
+			// optimization is enabled
+			for {
```
Why have an infinite for loop here? Couldn't we wait until the client is connected, then have something like 

```
<-kvstore.Client().Connected()
if !option.Config.K8sEventHandover {
    return
}
```


Then have the same logic as before?

tgraf replied:
Great question. I've added the loop so we can change the option at runtime. I excluded this in the scope of 1.5 but I want to add logic in 1.6 to enable this option (unless disabled by the user) as we pass 200 nodes in the cluster.

raybejjani replied:
I don't fully understand this, we're already in an infinite loop so we would already do this retrying. In that scenario, though, we would also bootstrap again via k8s. If we wanted to avoid that I'm not sure how this code does that, since it breaks the inner infinite loop, but would then fall down to `<-kvstore.Client().Disconnected()` and then loops again.
Since this is a scale issue, would allowing this watcher to connect ever make sense?

ianvernon replied:
@tgraf by "change the option at runtime" do you mean changing the configuration dynamically without restarting the cilium-agent? Is this possible via cobra / viper? Or rather that we would change the configuration dynamically if we scale up to over 200 nodes?

What about the scale-down case? I.e., what if we change from `option.Config.K8sEventHandover = false` to `option.Config>k8sEventHandover=true` - we will never enter back into this logic if we then set it back to `option.Config.K8sEventHandover = false` since we break out of the loop here. This means that once we rely on the kvstore, we always rely on it - which isn't a bad thing, just confirming that this is the desired behavior.

tgraf replied:
> by "change the option at runtime" do you mean changing the configuration dynamically without restarting the cilium-agent? Is this possible via cobra / viper? Or rather that we would change the configuration dynamically if we scale up to over 200 nodes?

Yes, this. Given that this seems non-obvious, I'll change this PR to not include this and convert it into a loop in the context of 1.6. It will be easier to review with all context.





> What about the scale-down case? I.e., what if we change from `option.Config.K8sEventHandover = false` to `option.Config>k8sEventHandover=true` - we will never enter back into this logic if we then set it back to `option.Config.K8sEventHandover = false` since we break out of the loop here. This means that once we rely on the kvstore, we always rely on it - which isn't a bad thing, just confirming that this is the desired behavior.

So far, the plan has been to never revert back unless cilium is restarted after scale-down has occured. Let's discuss when I propose the PR specifically. I think both options are correct.

tgraf replied:
Changed.

tgraf committed:
k8s: Fix leak of k8s controller on kvstore connect & disconnect

A go channel plus a k8s controller was leaked on each kvstore connect +
disconnect event as the minimal k8s event controller was replaced with a
full-scope controller.

Signed-off-by: Thomas Graf <thomas@cilium.io>

ianvernon commented on daemon/k8s_watcher.go:
```diff
@@ -841,6 +846,7 @@ func (d *Daemon) EnableK8sWatcher(queueSize uint) error {
 			// Create a new pod controller when we are disconnected with the
 			// kvstore
 			<-kvstore.Client().Disconnected()
+			close(isConnected)
```
Nice catch !
